### DISCUSSION TITLE
lmod issues since recent Lua update

### DISCUSSION THREAD

**Discussion by mathomp4** [Homebrew/discussions#218] [Everyday usage] [answered]
I believe this morning, a `brew  upgrade` [brought in Lua 5.4](https://github.com/Homebrew/homebrew-core/commits/master/Formula/lua.rb). But in doing so, my lmod (also from brew) install seems to have broken:
```console
❯ module
/Users/mathomp4/.homebrew/brew/opt/lua/bin/lua: version mismatch: app. needs 503.0, Lua core provides 504.0
stack traceback:
	[C]: in ?
	[C]: in function 'require'
	.../lmod/8.4.16/libexec/vendor/share/lua/5.3/posix/init.lua:23: in main chunk
	[C]: in function 'require'
	.../mathomp4/.homebrew/brew/Cellar/lmod/8.4.16/libexec/lmod:61: in main chunk
	[C]: in ?
```
From staring at it I guess it's from the `luaposix` rock? I do see in [the lmod formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/lmod.rb) some hardcoded 5.3. Maybe?

I'm a pretty weak Lua person, so I'm not sure how to fix this. Is it possible to "move back" to the old version? I really only need Lua for lmod.

ETA: Note: I made a discussion because I have no way of knowing if this is a reproducible error, thus I can't file an issue.

**Comment by carlocab** [comment]
Sorry about that. It passed CI tests and there were about a dozen other formulae complaining much more loudly about the update, so I didn't look too closely.

I'll have a look. Can you give me a complete set of repro steps? The output of `brew config` and `brew doctor` would also be helpful.

**Reply by carlocab** [reply]
Despite my PR, please still answer my questions regarding repro steps and the output of `brew config` and `brew doctor`. I need this in order to devise a test to make sure `lmod` doesn't break with future updates. Thank you.

**Reply by mathomp4** [reply]
Here's the output of the brew commands. First `config`:
```console
❯ brew config
HOMEBREW_VERSION: 2.6.0
ORIGIN: https://github.com/Homebrew/brew
HEAD: 1d5e354cc2ff048bd7161d95b3fa7f91dc9dd081
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: e24f5681e7e2a9667361686f2f4896f0c480af06
Core tap last commit: 3 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /Users/mathomp4/.homebrew/brew
HOMEBREW_REPOSITORY: /Users/mathomp4/.homebrew/brew
HOMEBREW_CELLAR: /Users/mathomp4/.homebrew/brew/Cellar
HOMEBREW_CASK_OPTS: ["--appdir=~/Applications"]
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.PEl3ZEs8dp/org.macosforge.xquartz:0
HOMEBREW_EDITOR: /Users/mathomp4/.homebrew/brew/bin/vim
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 12.0 build 1200
Git: 2.29.2 => /Users/mathomp4/.homebrew/brew/bin/git
Curl: 7.64.1 => /usr/bin/curl
Java: 11.0.9.1
macOS: 10.15.7-x86_64
CLT: 12.0.32.21
Xcode: 12.0.1
XQuartz: 2.7.11 => /opt/X11
```
and `doctor`:
```console
❯ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you an update run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/more/.


Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  guile@2

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause brews that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  lua@5.1
  md5sha1sum

Warning: Your Homebrew's prefix is not /usr/local.
Some of Homebrew's bottles (binary packages) can only be used with the default
prefix (/usr/local).
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
unsupported configuration.


Warning: Your Xcode (12.0.1) is outdated.
Please update to Xcode 12.2 (or delete it).
Xcode can be updated from the App Store.

```
Wow. Long.

Note: I have to do things outside of `/usr/local` because this isn't a personal Mac, so I have no rights to install anywhere else. But I have *never* encountered an issue so far!

**Reply by mathomp4** [reply]
As for how to reproduce, well, I guess you just need lmod as it currently is and try to run, say, `module` or `ml`:
```console
❯ module
/Users/mathomp4/.homebrew/brew/opt/lua/bin/lua: version mismatch: app. needs 503.0, Lua core provides 504.0
stack traceback:
	[C]: in ?
	[C]: in function 'require'
	.../lmod/8.4.16/libexec/vendor/share/lua/5.3/posix/init.lua:23: in main chunk
	[C]: in function 'require'
	.../mathomp4/.homebrew/brew/Cellar/lmod/8.4.16/libexec/lmod:61: in main chunk
	[C]: in ?
❯ ml
/Users/mathomp4/.homebrew/brew/opt/lua/bin/lua: version mismatch: app. needs 503.0, Lua core provides 504.0
stack traceback:
	[C]: in ?
	[C]: in function 'require'
	.../lmod/8.4.16/libexec/vendor/share/lua/5.3/posix/init.lua:23: in main chunk
	[C]: in function 'require'
	...athomp4/.homebrew/brew/Cellar/lmod/8.4.16/libexec/ml_cmd:62: in main chunk
	[C]: in ?
```

`ml` is a sort of "helper" function for `module`

**Reply by carlocab** [reply]
Does `module list` also give you an error?

**Reply by SMillerDev** [reply]
If I install it locally `source /usr/local/opt/lmod/init/profile` already gives an error. Could you install your fixed version locally and see if that works?

**Comment by carlocab** [comment, marked as answer]
I've proposed a fix here: https://github.com/Homebrew/homebrew-core/pull/66159

Does this work for you?

**Reply by mathomp4** [reply]
I sort of depend on the "magic" of `brew` so I haven't gotten too fancy with what I do. How would one go about testing that? 😄

**Reply by carlocab** [reply]
The simplest way is probably using `gh` or `hub` (I hope you have one of them installed and configured.)

Do the following, assuming you have `gh`:
1. `cd $(brew --repo homebrew/core)`
2. `gh pr checkout 66159`
3. `brew uninstall lmod && brew install -s lmod`
4. Check to see if `lmod` now works for you. (I don't use it, so I don't actually know what to do here.)

When you're finished, you can just go back to the `homebrew/core` directory and do `git checkout master`.

**Reply by mathomp4** [reply]
Well this is weird:
```console
❯ gh pr checkout 66159
Switched to branch 'lmod-fix'
❯ git branch
* lmod-fix
  master
❯ brew install -s lmod
==> Downloading https://github.com/keplerproject/luafilesystem/archive/v1_8_0.tar.gz
Already downloaded: /Users/mathomp4/Library/Caches/Homebrew/downloads/e38ef4e20ce21d56a5560c606a635836aeca5d5843f4c4afc73ac04ec4c8e407--luafilesystem-1_8_0.tar.gz
==> Downloading https://github.com/luaposix/luaposix/archive/v35.0.tar.gz
Already downloaded: /Users/mathomp4/Library/Caches/Homebrew/downloads/5adce844a92473ecdfc8af2f0bba615fded595f1399bf09d67d1ac3ee793b123--luaposix-35.0.tar.gz
==> Downloading https://github.com/TACC/Lmod/archive/8.4.16.tar.gz
Already downloaded: /Users/mathomp4/Library/Caches/Homebrew/downloads/bfdbf1ba54f0af1e6066e9b3f7f9f88fbd6525836ba0d29c4f6e7562b0dd1363--Lmod-8.4.16.tar.gz

==> luarocks make --tree=/Users/mathomp4/.homebrew/brew/Cellar/lmod/8.4.16/libexec/vendor
==> luarocks make --tree=/Users/mathomp4/.homebrew/brew/Cellar/lmod/8.4.16/libexec/vendor
==> ./configure --with-siteControlPrefix=yes --prefix=/Users/mathomp4/.homebrew/brew/Cellar/lmod/8.4.16
Last 15 lines from /Users/mathomp4/Library/Logs/Homebrew/lmod/03.configure:
checking for less... /usr/bin/less
checking for sha1sum... no
checking for gsha1sum... no
checking for shasum... /usr/bin/shasum
checking for tclsh... /usr/bin/tclsh
checking for gtr... no
checking for tr... /usr/bin/tr
checking for luac... /Users/mathomp4/.homebrew/brew/opt/lua/bin/luac
checking for lua... /Users/mathomp4/.homebrew/brew/opt/lua/bin/lua
checking for valid Lua version... 5.4
checking for lua modules: posix

Error: The follow lua module(s) are missing:  posix

You can not run Lmod without:  posix

Do not report this issue to Homebrew/brew or Homebrew/core!

These open issues may also help:
lmod: rebuild luarocks https://github.com/Homebrew/homebrew-core/pull/66159

❯ git branch
  lmod-fix
* master
```
For some reason, after it fails to build...I'm back on `master`. Is there an implicit `git switch master` in a `brew install` command?

**Reply by carlocab** [reply]
Not that I'm aware of... The build failure is odd though. It builds fine on my system, and on Homebrew CI.

**Reply by mathomp4** [reply]
Well, let me do it the "old fashioned way" and just copy-paste your changes over. I can always `git restore` the file afterwards

**Reply by mathomp4** [reply]
Oh. Interesting. I just tried again (before copy paste) and it's like `brew upgrade` ran automatically. Is there a way to prevent `brew install` from doing an implicit update?

**Reply by carlocab** [reply]
Yes. Do `HOMEBREW_NO_AUTO_UPDATE=1 brew install -s lmod`.

**Reply by mathomp4** [reply]
Well, I found out about `HOMEBREW_NO_AUTO_UPDATE=1` but now:
```console
❯ gh pr checkout 66159
From https://github.com/Homebrew/homebrew-core
 ! [rejected]              refs/pull/66159/head -> lmod-fix  (non-fast-forward)
exit status 1
```

**Reply by carlocab** [reply]
I think it's refusing to do that because you made changes to your local core repo before doing `gh pr checkout`. It might be better to just stay on `master` and do it the quick and dirty (copy and paste) way instead of trying to get `gh` to work...

**Reply by mathomp4** [reply]
@carlocab Okay. So, when I "hand-enter" all of your changes and run `HOMEBREW_NO_AUTO_UPDATE=1 brew install -s lmod`, I get lmod back and working! Huzzah!

**Reply by carlocab** [reply]
Great. The changes should be merged soon. Thanks for your patience.

**Comment by mathomp4** [comment]
Note, also, you might try consulting @rtmclay (creator and master magician of https://github.com/TACC/Lmod itself) if you are looking for a good way to detect an issue with lmod/luarocks/etc. He might know some lower-level thing that I am unaware of. (I've bothered him a few times when I've broken something in lmod.)

**Reply by carlocab** [reply]
AFAICT, the fix to your problem is easy: just rebuild `lmod`. (Unless there's some sort of subtle incompatibility I haven't picked up on.)

What bothers me is that CI didn't catch this breakage, when it really should've.

**Reply by mathomp4** [reply]
Hmm. So uninstall and reinstall `lmod`? Or is there is `brew rebuild` like command?

**Reply by carlocab** [reply]
Sorry, the fix on this end. On your end that won't quite work yet, as there needs to be an update to the `lmod` formula.

**Reply by mathomp4** [reply]
I did just try `brew reinstall lmod` but no joy. I'll try with `-s`...

**Reply by mathomp4** [reply]
And after your reply, I guess that's what did this:
```
checking for valid Lua version... 5.4
checking for lua modules: posix

Error: The follow lua module(s) are missing:  posix

You can not run Lmod without:  posix

Do not report this issue to Homebrew/brew or Homebrew/core!

These open issues may also help:
lmod: rebuild luarocks https://github.com/Homebrew/homebrew-core/pull/66159

Error: Your Xcode (12.0.1) is outdated.
Please update to Xcode 12.2 (or delete it).
Xcode can be updated from the App Store.


Error: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you an update run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/more/.
```
I'll wait for further instructions, and, I guess, update my Xcode as `brew` seems annoyed!
